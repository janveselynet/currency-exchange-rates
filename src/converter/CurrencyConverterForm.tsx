import React from 'react';
import styled from "styled-components";

const NumberInput = styled.input`
  margin-right: 10px;
  padding: 7px;
  border: 1px solid #d1d1d1;
  border-radius: 6px;
`;

const Select = styled.select`
  margin: 0 10px;
  padding: 7px;
  border-radius: 6px;
  border: 1px solid #d1d1d1;
`;

const Button = styled.button`
  margin-left: 10px;
  background-color: #2980b9;
  border-radius: 8px;
  border-style: none;
  box-sizing: border-box;
  color: #ffffff;
  cursor: pointer;
  display: inline-block;
  padding: 10px 16px;
  text-align: center;

  :hover,
  :focus {
    background-color: #5ca9db;
  }
`;

interface CurrencyConverterForm {
  availableCurrencyCodes: Array<string>;
  onFormSubmitted: (amountInCzk: number, targetedCurrencyCode: string) => void;
}

export default function CurrencyConverterForm({
  availableCurrencyCodes,
  onFormSubmitted,
}: CurrencyConverterForm): JSX.Element {
  const [amount, setAmount] = React.useState<string>('');
  const [currencyCode, setCurrencyCode] = React.useState<string>(availableCurrencyCodes[0]);

  const handleAmountChanged = (event: React.ChangeEvent<HTMLInputElement>): void => {
    setAmount(event.target.value);
  };

  const handleCurrencyCodeChanged = (event: React.ChangeEvent<HTMLSelectElement>): void => {
    setCurrencyCode(event.target.value);
  };

  const handleSubmit = (event: React.FormEvent<HTMLFormElement>): void => {
    event.preventDefault();

    const parsedAmount = parseFloat(amount) || 0;
    onFormSubmitted(parsedAmount, currencyCode);

    setAmount('');
  };

  return (
    <form onSubmit={handleSubmit}>
      <NumberInput
        id="amount"
        type="number"
        placeholder="Enter amount"
        value={amount}
        onChange={handleAmountChanged}
      />

      CZK to

      <Select
        name="currencyCode"
        value={currencyCode}
        onChange={handleCurrencyCodeChanged}
      >
        {availableCurrencyCodes.map((currencyCode, key) => (
          <option key={key} value={currencyCode}>{currencyCode}</option>
        ))}
      </Select>

      <Button>Convert</Button>
    </form>
  );
};
