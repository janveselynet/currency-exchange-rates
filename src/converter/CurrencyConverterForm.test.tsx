import { fireEvent, render, screen } from '@testing-library/react';
import React from 'react';
import CurrencyConverterForm from './CurrencyConverterForm';

describe('CurrencyConverterForm', () => {
  const availableCurrencyCodes = ['CZK', 'USD', 'EUR'];

  it('renders properly', () => {
    const { container } = render(
      <CurrencyConverterForm
        availableCurrencyCodes={availableCurrencyCodes}
        onFormSubmitted={jest.fn()}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('submits entered data when button is clicked', () => {
    const onFormSubmittedMock = jest.fn();

    render(
      <CurrencyConverterForm
        availableCurrencyCodes={availableCurrencyCodes}
        onFormSubmitted={onFormSubmittedMock}
      />
    );

    fireEvent.change(screen.getByRole('spinbutton'), { target: { value: '100' } });
    fireEvent.change(screen.getByRole('combobox'), { target: { value: 'USD' } });
    fireEvent.click(screen.getByRole('button'));

    expect(onFormSubmittedMock).toHaveBeenCalledWith(100, 'USD');
  });

  it('submits meaningful data if nothing is explicitly set', () => {
    const onFormSubmittedMock = jest.fn();

    render(
      <CurrencyConverterForm
        availableCurrencyCodes={availableCurrencyCodes}
        onFormSubmitted={onFormSubmittedMock}
      />
    );

    fireEvent.click(screen.getByRole('button'));

    expect(onFormSubmittedMock).toHaveBeenCalledWith(0, 'CZK');
  });

  it('clears amount input after submitting form', () => {
    render(
      <CurrencyConverterForm
        availableCurrencyCodes={availableCurrencyCodes}
        onFormSubmitted={jest.fn()}
      />
    );

    const amountInput = screen.getByRole('spinbutton');
    fireEvent.change(amountInput, { target: { value: '100' } });
    fireEvent.click(screen.getByRole('button'));

    expect(amountInput).toHaveValue(null);
  });
});
