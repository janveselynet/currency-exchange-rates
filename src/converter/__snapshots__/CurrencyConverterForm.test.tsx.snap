// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyConverterForm renders properly 1`] = `
<div>
  <form>
    <input
      class="sc-bcXHqe gkbLzw"
      id="amount"
      placeholder="Enter amount"
      type="number"
      value=""
    />
    CZK to
    <select
      class="sc-gswNZR edpAjZ"
      name="currencyCode"
    >
      <option
        value="CZK"
      >
        CZK
      </option>
      <option
        value="USD"
      >
        USD
      </option>
      <option
        value="EUR"
      >
        EUR
      </option>
    </select>
    <button
      class="sc-dkrFOg eqBICR"
    >
      Convert
    </button>
  </form>
</div>
`;
